src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs

```diff
@@ -0,0 +1,2255 @@
+            comp.VerifyDiagnostics(
+                // (4,25): error CS0569: 'Derived.Property': cannot override 'C.Property' because it is not supported by the language
+                //     public override int Property { set { throw null; } }
+                Diagnostic(ErrorCode.ERR_CantOverrideBogusMethod, "Property").WithArguments("Derived.Property", "C.Property").WithLocation(4, 25),
+                // (8,25): error CS0569: 'Derived2.Property': cannot override 'C.Property' because it is not supported by the language
+                //     public override int Property { init { throw null; } }
+                Diagnostic(ErrorCode.ERR_CantOverrideBogusMethod, "Property").WithArguments("Derived2.Property", "C.Property").WithLocation(8, 25)
+                );
+        }
+
+        [Fact]
+        public void ModReqOnGetAccessorReturnValue()
+        {
+            string il = @"
+.class public auto ansi beforefieldinit C extends System.Object
+{
+    .method public hidebysig specialname newslot virtual instance int32 modreq(System.Runtime.CompilerServices.IsExternalInit) get_Property () cil managed
+    {
+        IL_0000: ldnull
+        IL_0001: throw
+    }
+
+    .method public hidebysig specialname newslot virtual instance void set_Property ( int32 'value' ) cil managed
+    {
+        IL_0000: ldnull
+        IL_0001: throw
+    }
+
+    .method public hidebysig specialname rtspecialname instance void .ctor () cil managed
+    {
+        IL_0000: ldnull
+        IL_0001: throw
+    }
+
+    .property instance int32 Property()
+    {
+        .get instance int32 modreq(System.Runtime.CompilerServices.IsExternalInit) C::get_Property()
+        .set instance void C::set_Property(int32)
+    }
+}
+
+.class public auto ansi sealed beforefieldinit System.Runtime.CompilerServices.IsExternalInit extends System.Object
+{
+    .method public hidebysig specialname rtspecialname instance void .ctor () cil managed
+    {
+        IL_0000: ldnull
+        IL_0001: throw
+    }
+}
+";
+
+            string source = @"
+public class Derived : C
+{
+    public override int Property { get { throw null; } }
+}
+public class Derived2 : C
+{
+    public override int Property { get { throw null; } }
+}
+";
+
+            var reference = CreateMetadataReferenceFromIlSource(il);
+            var comp = CreateCompilation(source, references: new[] { reference }, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics();
+
+            var property = (PEPropertySymbol)comp.GlobalNamespace.GetMember("C.Property");
+            Assert.False(property.GetMethod.IsInitOnly);
+            Assert.False(property.SetMethod.IsInitOnly);
+        }
+
+        [Fact]
+        public void TestSyntaxFacts()
+        {
+            Assert.True(SyntaxFacts.IsAccessorDeclaration(SyntaxKind.InitAccessorDeclaration));
+            Assert.True(SyntaxFacts.IsAccessorDeclarationKeyword(SyntaxKind.InitKeyword));
+        }
+    }
+}
```

Comment:
Several general observations:
- It looks like majority of success cases are not executed and runtime behavior is not observed. I think we should do that at least for each unique scenario.
- There are no tests around ```base.```. We should cover success and error cases both for the init-only setters and for readonly fields.
- There are almost no direct tests for IsInitOnly API. We should cover at least all implementations that do not simply return false (covering those would be good as well, but I would say not absolutely required). Here is what I think must to have non trivial cases (both true/false results): PE, source, all leafs deriving from WrappedMethodSymbol (substituted, retargeting, etc.)

Reply:
There is only one test that I could find that tests for lack/presence of the modreq. It feels like we should have more. For example for scenarios involving overriding and implementing, including situations when there is a set/init mismatch.


---
In reply to: [415213347](https://github.com/dotnet/roslyn/pull/43275#discussion_r415213347) [](ancestors = 415213347)